File: android-app/app/src/main/java/com/redcode/im/androidapp/e2ee/E2eeIncomingMessageResolver.kt
package com.redcode.im.androidapp.e2ee

import com.redcode.im.androidapp.core.model.AuthSession
import com.redcode.im.androidapp.core.model.ChatMessage
import com.redcode.im.androidapp.data.chat.BackendChatMessage
import java.util.Base64
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class E2eeIncomingMessageException(message: String, cause: Throwable? = null) : Exception(message, cause)

interface IncomingChatMessageResolver {
    suspend fun resolve(
        message: BackendChatMessage,
        source: E2eeMessageSource,
        cachedMessage: ChatMessage? = null,
    ): ChatMessage

    suspend fun rememberResolved(message: ChatMessage) = Unit
}

object PlaintextIncomingMessageResolver : IncomingChatMessageResolver {
    override suspend fun resolve(
        message: BackendChatMessage,
        source: E2eeMessageSource,
        cachedMessage: ChatMessage?,
    ): ChatMessage = message.toDomain()
}

class E2eeIncomingMessageResolver(
    private val session: StateFlow<AuthSession?>,
    private val e2eeStatus: StateFlow<E2eeSessionStatus>,
    private val decryptor: E2eeIncomingDecryptor,
    private val deviceLabel: String,
) : IncomingChatMessageResolver {
    private val mutex = Mutex()
    private val resolvedMessages = LinkedHashMap<String, ChatMessage>()

    override suspend fun resolve(
        message: BackendChatMessage,
        source: E2eeMessageSource,
        cachedMessage: ChatMessage?,
    ): ChatMessage = mutex.withLock {
        if (message.isDeleted) return@withLock message.toDomain()
        val encryptedContent = message.encryptedContent
        val metadata = message.encryptionMetadata
        if (encryptedContent == null && metadata == null) return@withLock message.toDomain()
        if (encryptedContent.isNullOrBlank() || metadata == null) {
            throw E2eeIncomingMessageException("E2EE 消息密文与 metadata 不完整")
        }
        if (
            metadata.protocol != "mls" ||
            metadata.version != 1 ||
            metadata.epoch <= 0 ||
            metadata.senderDeviceId.isBlank() ||
            metadata.contentType != "application"
        ) {
            throw E2eeIncomingMessageException("E2EE 消息 metadata 无效")
        }
        val activeSession = session.value ?: throw E2eeIncomingMessageException("E2EE 消息解密需要登录态")
        when (val status = e2eeStatus.value) {
            E2eeSessionStatus.Plaintext, is E2eeSessionStatus.Ready -> Unit
            is E2eeSessionStatus.Blocked -> throw E2eeIncomingMessageException(status.message)
            E2eeSessionStatus.SignedOut -> throw E2eeIncomingMessageException("E2EE 会话未登录")
        }
        if (cachedMessage?.id == message.id) {
            val resolved =
                message.toDomain().copy(
                    text = cachedMessage.text,
                    parts = cachedMessage.parts,
                )
            remember(activeSession.user.id, resolved)
            return@withLock resolved
        }
        val cacheKey = cacheKey(activeSession.user.id, message.id)
        resolvedMessages[cacheKey]?.let { cached ->
            return@withLock message.toDomain().copy(text = cached.text, parts = cached.parts)
        }
        val ciphertext =
            try {
                Base64.getDecoder().decode(encryptedContent)
            } catch (error: IllegalArgumentException) {
                throw E2eeIncomingMessageException("E2EE 消息密文 Base64 无效", error)
            }
        if (ciphertext.isEmpty()) throw E2eeIncomingMessageException("E2EE 消息密文不能为空")
        val decrypted =
            try {
                decryptor.decryptIncoming(
                    accountId = activeSession.user.id,
                    deviceLabel = deviceLabel,
                    input = E2eeIncomingMessage(message.id, message.roomId, ciphertext, source = source),
                    token = activeSession.tokens.accessToken,
                )
            } catch (error: E2eeIncomingMessageException) {
                throw error
            } catch (error: Exception) {
                throw E2eeIncomingMessageException("E2EE 消息解密失败", error)
            }
        if (
            decrypted.messageId != message.id ||
            decrypted.roomId != message.roomId ||
            decrypted.epoch != metadata.epoch ||
            !decrypted.encrypted
        ) {
            throw E2eeIncomingMessageException("E2EE 解密结果与消息不匹配")
        }
        val resolved = message.toDomain().copy(text = decrypted.text)
        remember(activeSession.user.id, resolved)
        resolved
    }

    override suspend fun rememberResolved(message: ChatMessage) = mutex.withLock {
        val accountId = session.value?.user?.id ?: return@withLock
        remember(accountId, message)
    }

    private fun remember(accountId: String, message: ChatMessage) {
        resolvedMessages[cacheKey(accountId, message.id)] = message
        while (resolvedMessages.size > MAX_RESOLVED_MESSAGES) {
            resolvedMessages.remove(resolvedMessages.keys.first())
        }
    }

    private fun cacheKey(accountId: String, messageId: String) = "$accountId:$messageId"

    private companion object {
        const val MAX_RESOLVED_MESSAGES = 2_000
    }
}

File: ios-app/Sources/RedCodeFeatures/E2eeIncomingMessageResolver.swift
import Foundation
import RedCodeCore
import RedCodeNetworking

public struct E2eeIncomingMessageResolverError: Error, Equatable, LocalizedError, Sendable {
    public let message: String

    public init(_ message: String) {
        self.message = message
    }

    public var errorDescription: String? { message }
}

public protocol E2eeIncomingMessageDecrypting: Sendable {
    func decryptIncoming(
        accountID: String,
        deviceLabel: String,
        input: E2eeIncomingMessage,
        token: String
    ) async throws -> E2eeDecryptedMessage
}

extension E2eeDirectMessageCoordinator: E2eeIncomingMessageDecrypting {}

@MainActor
public protocol E2eeSessionStatusProviding: AnyObject {
    var status: E2eeSessionStatus { get }
}

extension E2eeSessionLifecycle: E2eeSessionStatusProviding {}

@MainActor
public protocol IncomingChatMessageResolving: AnyObject {
    func resolve(
        _ message: ChatMessage,
        source: E2eeMessageSource,
        accountID: String,
        token: String,
        cachedMessage: ChatMessage?
    ) async throws -> ChatMessage

    func rememberResolved(_ message: ChatMessage, accountID: String)
}

@MainActor
public final class PlaintextIncomingMessageResolver: IncomingChatMessageResolving {
    public init() {}

    public func resolve(
        _ message: ChatMessage,
        source: E2eeMessageSource,
        accountID: String,
        token: String,
        cachedMessage: ChatMessage?
    ) async throws -> ChatMessage {
        message
    }

    public func rememberResolved(_ message: ChatMessage, accountID: String) {}
}

@MainActor
public final class E2eeIncomingMessageResolver: IncomingChatMessageResolving {
    private let sessionStatus: any E2eeSessionStatusProviding
    private let decryptor: any E2eeIncomingMessageDecrypting
    private let deviceLabel: String
    private var resolvedMessages: [String: ChatMessage] = [:]
    private var resolvedMessageOrder: [String] = []

    public init(
        sessionStatus: any E2eeSessionStatusProviding,
        decryptor: any E2eeIncomingMessageDecrypting,
        deviceLabel: String
    ) {
        self.sessionStatus = sessionStatus
        self.decryptor = decryptor
        self.deviceLabel = deviceLabel
    }

    public func resolve(
        _ message: ChatMessage,
        source: E2eeMessageSource,
        accountID: String,
        token: String,
        cachedMessage: ChatMessage? = nil
    ) async throws -> ChatMessage {
        if message.isDeleted { return message }

        let encryptedContent = message.encryptedContent
        let metadata = message.encryptionMetadata
        if encryptedContent == nil, metadata == nil { return message }
        guard let encryptedContent, !encryptedContent.isEmpty, let metadata else {
            throw E2eeIncomingMessageResolverError("E2EE 消息密文与 metadata 不完整")
        }
        guard metadata.protocolName == "mls",
              metadata.version == 1,
              metadata.epoch > 0,
              !metadata.senderDeviceID.isEmpty,
              metadata.contentType == "application" else {
            throw E2eeIncomingMessageResolverError("E2EE 消息 metadata 无效")
        }

        try validateSession(accountID: accountID)

        if let cachedMessage,
           cachedMessage.id == message.id,
           isReusableResolvedMessage(cachedMessage) {
            let resolved = message.replacingDecryptedContent(
                cachedMessage.content,
                parts: cachedMessage.parts
            )
            rememberResolved(resolved, accountID: accountID)
            return resolved
        }

        let key = cacheKey(accountID: accountID, messageID: message.id)
        if let cached = resolvedMessages[key], isReusableResolvedMessage(cached) {
            return message.replacingDecryptedContent(cached.content, parts: cached.parts)
        }

        guard let ciphertext = Data(base64Encoded: encryptedContent), !ciphertext.isEmpty else {
            throw E2eeIncomingMessageResolverError("E2EE 消息密文 Base64 无效")
        }

        let decrypted: E2eeDecryptedMessage
        do {
            decrypted = try await decryptor.decryptIncoming(
                accountID: accountID,
                deviceLabel: deviceLabel,
                input: E2eeIncomingMessage(
                    messageID: message.id,
                    roomID: message.roomID,
                    ciphertext: ciphertext,
                    source: source
                ),
                token: token
            )
        } catch {
            throw E2eeIncomingMessageResolverError("E2EE 消息解密失败")
        }

        guard decrypted.messageID == message.id,
              decrypted.roomID == message.roomID,
              decrypted.epoch == metadata.epoch,
              decrypted.encrypted else {
            throw E2eeIncomingMessageResolverError("E2EE 解密结果与消息不匹配")
        }

        let parts = decrypted.attachmentParts.map { part in
            ChatMessagePart(
                position: Int(part.partPosition),
                partType: messageType(for: part.mimeType),
                attachment: ChatMessageAttachment(
                    key: part.objectKey,
                    name: part.name,
                    mimeType: part.mimeType,
                    size: part.size
                )
            )
        }
        let resolved = message.replacingDecryptedContent(decrypted.text, parts: parts)
        rememberResolved(resolved, accountID: accountID)
        return resolved
    }

    public func rememberResolved(_ message: ChatMessage, accountID: String) {
        let key = cacheKey(accountID: accountID, messageID: message.id)
        if resolvedMessages[key] == nil {
            resolvedMessageOrder.append(key)
        }
        resolvedMessages[key] = message
        while resolvedMessageOrder.count > 2_000 {
            let oldest = resolvedMessageOrder.removeFirst()
            resolvedMessages.removeValue(forKey: oldest)
        }
    }

    private func validateSession(accountID: String) throws {
        switch sessionStatus.status {
        case .plaintext:
            return
        case .ready(let activeAccountID, _):
            guard activeAccountID == accountID else {
                throw E2eeIncomingMessageResolverError("E2EE 消息账号与当前会话不匹配")
            }
        case .blocked(let message):
            throw E2eeIncomingMessageResolverError(message)
        case .signedOut:
            throw E2eeIncomingMessageResolverError("E2EE 会话未登录")
        }
    }

    private func cacheKey(accountID: String, messageID: String) -> String {
        "\(accountID):\(messageID)"
    }

    private func isReusableResolvedMessage(_ message: ChatMessage) -> Bool {
        !message.content.isEmpty || !message.parts.isEmpty
    }

    private func messageType(for mimeType: String) -> ChatMessageType {
        if mimeType.hasPrefix("image/") { return .image }
        if mimeType.hasPrefix("video/") { return .video }
        if mimeType.hasPrefix("audio/") { return .audio }
        return .file
    }
}

private extension ChatMessage {
    func replacingDecryptedContent(_ content: String, parts: [ChatMessagePart]) -> ChatMessage {
        ChatMessage(
            id: id,
            roomID: roomID,
            senderID: senderID,
            senderName: senderName,
            content: content,
            encryptedContent: encryptedContent,
            encryptionMetadata: encryptionMetadata,
            messageType: messageType,
            status: status,
            timestamp: timestamp,
            isDeleted: isDeleted,
            isPinned: isPinned,
            pinnedAt: pinnedAt,
            pinnedBy: pinnedBy,
            quotedMessage: quotedMessage,
            parts: parts,
            attachments: parts.compactMap(\.attachment),
            reactions: reactions
        )
    }
}
